beluga/include/beluga/algorithm/mcl_core.hpp

```diff
@@ -0,0 +1,187 @@
+// Copyright 2024 Ekumen, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef BELUGA_ALGORITHM_MCL_CORE_HPP
+#define BELUGA_ALGORITHM_MCL_CORE_HPP
+
+#include <execution>
+#include <optional>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include <range/v3/range/concepts.hpp>
+#include <range/v3/view/any_view.hpp>
+#include <range/v3/view/take_exactly.hpp>
+#include "beluga/actions.hpp"
+#include "beluga/algorithm/estimation.hpp"
+#include "beluga/containers/circular_array.hpp"
+#include "beluga/containers/tuple_vector.hpp"
+#include "beluga/policies/every_n.hpp"
+#include "beluga/policies/on_effective_size_drop.hpp"
+#include "beluga/policies/on_motion.hpp"
+#include "beluga/random/multivariate_normal_distribution.hpp"
+#include "beluga/views/sample.hpp"
+
+namespace beluga {
+
+/// Struct containing parameters for the Monte Carlo Localization (MCL) implementation.
+struct MclParams {
+  /// Min distance in meters between updates.
+  double update_min_d = 0.25;
+  /// Min angular distance in radians between updates.
+  double update_min_a = 0.2;
+  /// Filter iterations interval at which a resampling will happen, unitless.
+  std::size_t resample_interval = 1UL;
+  /// Whether to use selective resampling or not.
+  bool selective_resampling = false;
+  /// Number of particles in the filter at any point in time.
+  std::size_t num_particles = 1000UL;
+};
+
+/// Implementation of the Monte Carlo Localization (MCL) algorithm.
+/**
+ * \tparam MotionModel Class representing a motion model. Must satisfy \ref MotionModelPage.
+ * \tparam SensorModel Class representing a sensor model. Must satisfy \ref SensorModelPage.
+ * \tparam WeightT Type to represent a weight of a particle.
+ * \tparam ParticleType Full particle type, containing state, weight and possibly
+ * other information .
+ * \tparam ExecutionPolicy Execution policy for particles processing.
+ */
+template <
+    class MotionModel,
+    class SensorModel,
+    typename WeightT = beluga::Weight,
+    class ParticleType = std::tuple<typename SensorModel::state_type, WeightT>,
```

Comment:
@serraramiro1 @nahueespinosa meta: moving forward, this looks unnecessarily restrictive. `SensorModel` could work with multiple state types. Same can be said about `MotionModel`.

Reply:
I'm wondering if we should couple the particle type and the models like these. Models could work with multiple types at the same time (e.g. through generic lambdas), so sparing the user from explicitly setting the particle state type on class instantiation seems more trouble than it's worth.